core/src/kvstore/writelog.rs

```diff
@@ -1,105 +1,336 @@
+#[derive(Debug)]
+struct Logger {
+    writer: Box<LogWriter>,
+}
 
-            let key = Key::read(&buf[0..24]);
-            let ts = BigEndian::read_i64(&buf[24..32]);
-            let exists = buf[32] != 0;
+impl Logger {
+    fn memory() -> Self {
+        Logger {
+            writer: Box::new(CRCWriter::new(vec![], BLOCK_SIZE)),
+        }
+    }
 
-            let val = if exists {
-                Some(buf[33..].to_vec())
-            } else {
-                None
-            };
-            let value = Value { ts, val };
+    fn disk(file: File) -> Self {
+        Logger {
+            writer: Box::new(CRCWriter::new(file, BLOCK_SIZE)),
+        }
+    }
+}
+
+impl LogWriter for CRCWriter<Vec<u8>> {
+    fn sync(&mut self, _: bool) -> Result<()> {
+        Ok(self.flush()?)
+    }
 
-            table.insert(key, value);
+    fn mmap(&self) -> Result<Mmap> {
+        let mut map = memmap::MmapMut::map_anon(self.get_ref().len())?;
+        (&mut map[..]).copy_from_slice(self.get_ref());
+        Ok(map.make_read_only()?)
+    }
+}
+
+impl LogWriter for CRCWriter<File> {
+    fn sync(&mut self, fsync: bool) -> Result<()> {
+        self.flush()?;
+
+        let file = self.get_mut();
+        if fsync {
+            file.sync_all()?;
+        } else {
+            file.sync_data()?;
         }
 
-        Ok(table)
+        Ok(())
+    }
+
+    fn mmap(&self) -> Result<Mmap> {
+        let map = unsafe { Mmap::map(self.get_ref())? };
+        Ok(map)
+    }
+}
+
+fn log(logger: &mut Logger, key: &Key, commit: i64, data: Option<&[u8]>) -> Result<()> {
+    let wtr = &mut logger.writer;
+    write_value(wtr, key, commit, data)?;
+
+    Ok(())
+}
+
+fn sync(logger: &mut Logger, sync_all: bool) -> Result<()> {
+    let writer = &mut logger.writer;
+
+    writer.sync(sync_all)?;
+
+    Ok(())
+}
+
+#[inline]
+fn file_opts() -> fs::OpenOptions {
+    let mut opts = fs::OpenOptions::new();
+    opts.read(true).write(true).create(true);
+    opts
+}
+
+fn read_log(wal: &[u8]) -> Result<BTreeMap<Key, Value>> {
+    let mut map = BTreeMap::new();
+    if wal.len() <= 8 + 24 + 8 + 1 {
+        return Ok(map);
     }
+
+    let mut rdr = CRCReader::new(wal, BLOCK_SIZE);
+
+    while let Ok((key, val)) = read_value(&mut rdr) {
+        map.insert(key, val);
+    }
+
+    Ok(map)
 }
 
 #[inline]
-fn log_to_buffer(buf: &mut [u8], rec_len: u64, key: &Key, ts: i64, val: &[u8]) {
-    BigEndian::write_u64(&mut buf[..8], rec_len);
-    (&mut buf[8..32]).copy_from_slice(&key.0);
-    BigEndian::write_i64(&mut buf[32..40], ts);
-    buf[40] = (!val.is_empty()) as u8;
-    (&mut buf[41..]).copy_from_slice(val);
+fn write_value<W: Write>(wtr: &mut W, key: &Key, commit: i64, data: Option<&[u8]>) -> Result<()> {
+    let len = 24 + 8 + 1 + data.map(<[u8]>::len).unwrap_or(0);
+
+    wtr.write_u64::<BigEndian>(len as u64)?;
+    wtr.write_all(&key.0)?;
+    wtr.write_i64::<BigEndian>(commit)?;
+
+    match data {
+        Some(data) => {
```

Comment:
this looks a lot like serialize(), can we use Rust here instead of hand-rolling?

Reply:
Yes. I've made an issue for this here: https://github.com/solana-labs/solana/issues/3313.

I figure it should probably be another PR